diff --git a/aapg/gen_random_program.py b/aapg/gen_random_program.py
index f112a44..45a8f62 100644
--- a/aapg/gen_random_program.py
+++ b/aapg/gen_random_program.py
@@ -3409,7 +3409,7 @@ def run(args, index):
       lines = output_file.readlines()
 
     
-    lines[4] = '#include "{template}"\n#define TEST_NAME {test_name}\n'.format(template=line_add,test_name=name_add)
+    lines[4] = '#include \"aapg_common.h\"\n#define TEST_NAME {test_name}\n'.format(template=line_add,test_name=name_add)
 
     with open(output_file_path, "w") as outfile:
       outfile.write("".join(lines))
diff --git a/aapg/isa_funcs.py b/aapg/isa_funcs.py
index 6ff3c0e..17eab31 100644
--- a/aapg/isa_funcs.py
+++ b/aapg/isa_funcs.py
@@ -453,6 +453,20 @@ inst_store_by_set = {
     ]
 }
 
+inst_store_by_set.update({
+    'rv32zmmul' : [entry for entry in inst_store_by_set['rv32m'] if entry.startswith('mul')],
+})
+
+# custom isa
+inst_store_by_set.update({
+    'rv32simd.dot' : [
+        'dot16  rd rs1 rs2 31..25=0x06 14..12=0 6..2=0x02 1..0=3',
+        'dot8   rd rs1 rs2 31..25=0x07 14..12=0 6..2=0x02 1..0=3',
+        #'dot4   rd rs1 rs2 31..25=0x46 14..12=0 6..2=0x02 1..0=3',
+        #'dot2   rd rs1 rs2 31..25=0x47 14..12=0 6..2=0x02 1..0=3',
+    ],
+})
+
 comp_insts_subs = {
     'addi4spn': 'addi',
     'c.fld': 'fld',
diff --git a/aapg/main.py b/aapg/main.py
index 1c3303a..764194f 100644
--- a/aapg/main.py
+++ b/aapg/main.py
@@ -95,7 +95,7 @@ def gen(num_programs,config_file,asm_name,setup_dir,output_dir,arch,seed,linker_
     logger = logging.getLogger()
     logger.handlers = []
     ch = logging.StreamHandler()
-    ch.setFormatter(aapg.utils.ColoredFormatter())
+    #ch.setFormatter(aapg.utils.ColoredFormatter())
     logger.addHandler(ch)
     logger.info("aapg started")
     logger.info(VERSION)
@@ -159,7 +159,7 @@ def setup(setup_dir):
     logger = logging.getLogger()
     logger.handlers = []
     ch = logging.StreamHandler()
-    ch.setFormatter(aapg.utils.ColoredFormatter())
+    #ch.setFormatter(aapg.utils.ColoredFormatter())
     logger.addHandler(ch)
     logger.info("aapg started")
     logger.info(VERSION)
diff --git a/setup.py b/setup.py
index 094f585..7a953b0 100644
--- a/setup.py
+++ b/setup.py
@@ -50,3 +50,16 @@ setup(
     tests_require=test_requirements,
     zip_safe=False,
 )
+#setup(
+#    name='aapg',
+#    version='2.4.2',
+#    description="RISC-V AAPG",
+#    long_description=' AAPG \n\n',
+#    package_dir={'aapg': 'aapg'},
+#    entry_points={
+#        'console_scripts': ['aapg=aapg.main:cli'],
+#    },
+#    include_package_data=True,
+#    keywords='aapg',
+#    test_suite='tests',
+#)
