diff --git a/include/opcode/riscv-opc.h b/include/opcode/riscv-opc.h
index e77b49a6298..3e51c6785b3 100644
--- a/include/opcode/riscv-opc.h
+++ b/include/opcode/riscv-opc.h
@@ -3473,6 +3473,42 @@
 #define CSR_VL 0xc20
 #define CSR_VTYPE 0xc21
 #define CSR_VLENB 0xc22
+
+/* custom extension */
+#define MASK_CUSTOM_REG 0xfe00707f
+#define MASK_CUSTOM_IMM 0xfff0707f
+
+#define MATCH_ADD16   0xb
+#define MATCH_ADD8    0x200b
+#define MATCH_SUB16   0x400b
+#define MATCH_SUB8    0x600b
+
+#define MATCH_WMUL16   0x400000b
+#define MATCH_WMUL16U  0x400100b
+#define MATCH_WMUL8    0x400200b
+#define MATCH_WMUL8U   0x400300b
+
+#define MATCH_DOT16   0x600000b
+#define MATCH_DOT16U  0x600100b
+#define MATCH_DOT8    0x600200b
+#define MATCH_DOT8U   0x600300b
+#define MATCH_DOT4    0x600400b
+#define MATCH_DOT4U   0x600500b
+#define MATCH_DOT2    0x600600b
+#define MATCH_DOT2U   0x600700b
+
+#define MATCH_WIDEN16  0x4000000b
+#define MATCH_WIDEN16U 0x4000100b
+#define MATCH_WIDEN8   0x4000200b
+#define MATCH_WIDEN8U  0x4000300b
+#define MATCH_WIDEN4   0x4000400b
+#define MATCH_WIDEN4U  0x4000500b
+#define MATCH_WIDEN2   0x4000600b
+#define MATCH_WIDEN2U  0x4000700b
+
+#define MATCH_SCP_LCL 0xfe00000b
+#define MATCH_SCP_REL 0xfe00100b
+
 #endif /* RISCV_ENCODING_H */
 #ifdef DECLARE_INSN
 DECLARE_INSN(slli_rv32, MATCH_SLLI_RV32, MASK_SLLI_RV32)
diff --git a/opcodes/riscv-opc.c b/opcodes/riscv-opc.c
index fdd05ac75dc..8e025c6db4c 100644
--- a/opcodes/riscv-opc.c
+++ b/opcodes/riscv-opc.c
@@ -347,6 +347,38 @@ const struct riscv_opcode riscv_opcodes[] =
 {
 /* name, xlen, isa, operands, match, mask, match_func, pinfo.  */
 
+/* custom extension */
+{"add16",       0, INSN_CLASS_I, "d,s,t",     MATCH_ADD16, MASK_CUSTOM_REG, match_opcode, 0 },
+{"add8",        0, INSN_CLASS_I, "d,s,t",     MATCH_ADD8, MASK_CUSTOM_REG, match_opcode, 0 },
+{"sub16",       0, INSN_CLASS_I, "d,s,t",     MATCH_SUB16, MASK_CUSTOM_REG, match_opcode, 0 },
+{"sub8",        0, INSN_CLASS_I, "d,s,t",     MATCH_SUB8, MASK_CUSTOM_REG, match_opcode, 0 },
+
+{"wmul16",      0, INSN_CLASS_I, "d,s,t",     MATCH_WMUL16, MASK_CUSTOM_REG, match_opcode, 0 },
+{"wmul16u",     0, INSN_CLASS_I, "d,s,t",     MATCH_WMUL16U, MASK_CUSTOM_REG, match_opcode, 0 },
+{"wmul8",       0, INSN_CLASS_I, "d,s,t",     MATCH_WMUL8, MASK_CUSTOM_REG, match_opcode, 0 },
+{"wmul8u",      0, INSN_CLASS_I, "d,s,t",     MATCH_WMUL8U, MASK_CUSTOM_REG, match_opcode, 0 },
+
+{"dot16",       0, INSN_CLASS_I, "d,s,t",     MATCH_DOT16, MASK_CUSTOM_REG, match_opcode, 0 },
+{"dot16u",      0, INSN_CLASS_I, "d,s,t",     MATCH_DOT16U, MASK_CUSTOM_REG, match_opcode, 0 },
+{"dot8",        0, INSN_CLASS_I, "d,s,t",     MATCH_DOT8, MASK_CUSTOM_REG, match_opcode, 0 },
+{"dot8u",       0, INSN_CLASS_I, "d,s,t",     MATCH_DOT8U, MASK_CUSTOM_REG, match_opcode, 0 },
+{"dot4",        0, INSN_CLASS_I, "d,s,t",     MATCH_DOT4, MASK_CUSTOM_REG, match_opcode, 0 },
+{"dot4u",       0, INSN_CLASS_I, "d,s,t",     MATCH_DOT4U, MASK_CUSTOM_REG, match_opcode, 0 },
+{"dot2",        0, INSN_CLASS_I, "d,s,t",     MATCH_DOT2, MASK_CUSTOM_REG, match_opcode, 0 },
+{"dot2u",       0, INSN_CLASS_I, "d,s,t",     MATCH_DOT2U, MASK_CUSTOM_REG, match_opcode, 0 },
+
+{"widen16",     0, INSN_CLASS_I, "d,s",       MATCH_WIDEN16, MASK_CUSTOM_IMM, match_opcode, 0 },
+{"widen16u",    0, INSN_CLASS_I, "d,s",       MATCH_WIDEN16U, MASK_CUSTOM_IMM, match_opcode, 0 },
+{"widen8",      0, INSN_CLASS_I, "d,s",       MATCH_WIDEN8, MASK_CUSTOM_IMM, match_opcode, 0 },
+{"widen8u",     0, INSN_CLASS_I, "d,s",       MATCH_WIDEN8U, MASK_CUSTOM_IMM, match_opcode, 0 },
+{"widen4",      0, INSN_CLASS_I, "d,s",       MATCH_WIDEN4, MASK_CUSTOM_IMM, match_opcode, 0 },
+{"widen4u",     0, INSN_CLASS_I, "d,s",       MATCH_WIDEN4U, MASK_CUSTOM_IMM, match_opcode, 0 },
+{"widen2",      0, INSN_CLASS_I, "d,s",       MATCH_WIDEN2, MASK_CUSTOM_IMM, match_opcode, 0 },
+{"widen2u",     0, INSN_CLASS_I, "d,s",       MATCH_WIDEN2U, MASK_CUSTOM_IMM, match_opcode, 0 },
+
+{"scp.lcl",     0, INSN_CLASS_I, "d,s",       MATCH_SCP_LCL, MASK_CUSTOM_IMM, match_opcode, 0 },
+{"scp.rel",     0, INSN_CLASS_I, "d,s",       MATCH_SCP_REL, MASK_CUSTOM_IMM, match_opcode, 0 },
+
 /* Standard hints.  */
 {"prefetch.i",  0, INSN_CLASS_ZICBOP, "Wif(s)", MATCH_PREFETCH_I, MASK_PREFETCH_I, match_opcode, 0 },
 {"prefetch.r",  0, INSN_CLASS_ZICBOP, "Wif(s)", MATCH_PREFETCH_R, MASK_PREFETCH_R, match_opcode, 0 },
