diff --git a/framework/src/act/data/Gemfile.lock b/framework/src/act/data/Gemfile.lock
index 146ae9b2..b46010e2 100644
--- a/framework/src/act/data/Gemfile.lock
+++ b/framework/src/act/data/Gemfile.lock
@@ -223,7 +223,7 @@ CHECKSUMS
   awesome_print (1.9.2) sha256=e99b32b704acff16d768b3468680793ced40bfdc4537eb07e06a4be11133786e
   base64 (0.3.0) sha256=27337aeabad6ffae05c265c450490628ef3ebd4b67be58257393227588f5a97b
   bigdecimal (4.1.2) sha256=53d217666027eab4280346fba98e7d5b66baaae1b9c3c1c0ffe89d48188a3fbd
-  bundler (4.0.13) sha256=19f08be7f27022cf0b89f27da0b044ae075e8270a9ef44ad248a932614e1ca3b
+  bundler (4.0.14) sha256=d09a0a965cf772266a7e49e83610be7c2f4e49e61134c42a56804bb383cc24b8
   commander (5.0.0) sha256=22ca42866ce5ed1a7f3e9a9c831a23d03b75b3743c7ad0302c3cbcd9eb5f257d
   concurrent-ruby (1.3.6) sha256=6b56837e1e7e5292f9864f34b69c5a2cbc75c0cf5338f1ce9903d10fa762d5ab
   connection_pool (3.0.2) sha256=33fff5ba71a12d2aa26cb72b1db8bba2a1a01823559fb01d29eb74c286e62e0a
diff --git a/tests/env/rvtest_setup.h b/tests/env/rvtest_setup.h
index ab0433c4..37e92321 100644
--- a/tests/env/rvtest_setup.h
+++ b/tests/env/rvtest_setup.h
@@ -426,7 +426,7 @@
 
       // Disable interrupts
       csrw mie, zero
-      csrw mip, zero
+      //csrw mip, zero
 
       // disable trap delegation
       #ifdef S_SUPPORTED
diff --git a/tests/env/sail_macros.h b/tests/env/sail_macros.h
index 1acb471d..91b2618c 100644
--- a/tests/env/sail_macros.h
+++ b/tests/env/sail_macros.h
@@ -89,11 +89,11 @@
 
 ##### Machine Timer #####
 
-#undef RVMODEL_MTIMECMP_ADDRESS
-#define RVMODEL_MTIMECMP_ADDRESS  0x02004000  /* Address of mtimecmp CSR */
+//#undef RVMODEL_MTIMECMP_ADDRESS
+//#define RVMODEL_MTIMECMP_ADDRESS  0x02004000  /* Address of mtimecmp CSR */
 
-#undef RVMODEL_MTIME_ADDRESS
-#define RVMODEL_MTIME_ADDRESS  0x0200BFF8  /* Address of mtime CSR */
+//#undef RVMODEL_MTIME_ADDRESS
+//#define RVMODEL_MTIME_ADDRESS  0x0200BFF8  /* Address of mtime CSR */
 
 ##### Machine Interrupts #####
 
